File: packages/prompt-core-python/src/agentmark/prompt_core/template_engines/templatedx.py
"""TemplateDX-based template engine for prompt-core."""

import asyncio
from typing import Any

import yaml

from ..schemas import (
    AgentmarkConfigSchema,
    ImageConfigSchema,
    ObjectConfigSchema,
    SpeechConfigSchema,
    TextConfigSchema,
)
from ..types import PromptKind, RichChatMessage
from .instances import (
    ASSISTANT,
    EXTRACT_TEXT_PROMISES_KEY,
    IMAGE_PROMPT,
    SPEECH_PROMPT,
    SYSTEM,
    USER,
    determine_prompt_type,
    get_templatedx_instance,
)

# Type alias for AST nodes
Node = dict[str, Any]


def get_front_matter(tree: Node) -> dict[str, Any]:
    """Extract frontmatter from AST.

    Args:
        tree: Root AST node

    Returns:
        Parsed YAML frontmatter as dict
    """
    children = tree.get("children", [])
    for child in children:
        if child.get("type") == "yaml":
            value = child.get("value", "")
            result = yaml.safe_load(value)
            return result if isinstance(result, dict) else {}
    return {}


class TemplateDXTemplateEngine:
    """TemplateDX-based template engine for prompt-core."""

    async def compile(
        self,
        template: Any,
        props: dict[str, Any] | None = None,
    ) -> AgentmarkConfigSchema:
        """Compile a template with props.

        Args:
            template: Pre-parsed MDX AST
            props: Props to pass to the template

        Returns:
            Compiled prompt configuration
        """
        return await get_raw_config(ast=template, props=props or {})


async def get_raw_config(
    ast: Node,
    props: dict[str, Any],
) -> AgentmarkConfigSchema:
    """Extract configuration from AST with props applied.

    Args:
        ast: Pre-parsed MDX AST
        props: Props to pass to the template

    Returns:
        Compiled prompt configuration

    Raises:
        ValueError: If no valid config found in frontmatter
    """
    front_matter = get_front_matter(ast)
    prompt_type = determine_prompt_type(front_matter)

    templatedx_instance = get_templatedx_instance(prompt_type)
    shared: dict[str, Any] = {}
    await templatedx_instance.transform(ast, props, shared)

    # Collect extracted fields from coroutines
    extract_coros: list[Any] = shared.get(EXTRACT_TEXT_PROMISES_KEY, [])
    extracted_fields: list[dict[str, Any]] = list(await asyncio.gather(*extract_coros))

    # Build config based on type
    name: str = front_matter.get("name", "")
    speech_settings = front_matter.get("speech_config")
    image_settings = front_matter.get("image_config")
    object_settings = front_matter.get("object_config")
    text_settings = front_matter.get("text_config")
    test_settings = front_matter.get("test_settings")
    agentmark_meta = front_matter.get("agentmark_meta")

    # Determine config type and build response
    if speech_settings:
        return _build_speech_config(
            name, extracted_fields, speech_settings, test_settings, agentmark_meta
        )
    elif image_settings:
        return _build_image_config(
            name, extracted_fields, image_settings, test_settings, agentmark_meta
        )
    elif object_settings:
        return _build_object_config(
            name, extracted_fields, object_settings, test_settings, agentmark_meta
        )
    elif text_settings:
        return _build_text_config(
            name, extracted_fields, text_settings, test_settings, agentmark_meta
        )

    raise ValueError("No valid config found in frontmatter.")


def _get_messages(
    extracted_fields: list[dict[str, Any]], config_type: PromptKind
) -> list[RichChatMessage]:
    """Convert extracted fields to message list.

    Args:
        extracted_fields: List of extracted field dicts with 'name' and 'content'
        config_type: The prompt type for error messages

    Returns:
        List of rich chat messages

    Raises:
        ValueError: If system message is not first or invalid role tag
    """
    messages: list[RichChatMessage] = []
    role_tags = {USER, SYSTEM, ASSISTANT}

    for i, field in enumerate(extracted_fields):
        field_name = field["name"]

        if i != 0 and field_name == SYSTEM:
            raise ValueError(f"System message may only be the first message: {field['content']}")

        if field_name not in role_tags:
            raise ValueError(f'Invalid role tag: "{field_name}" in config type: {config_type}.')

        role_map = {USER: "user", ASSISTANT: "assistant", SYSTEM: "system"}
        role = role_map[field_name]
        messages.append({"role": role, "content": field["content"]})  # type: ignore[typeddict-item]

    return messages


def _get_prompt(
    tag_name: str,
    extracted_fields: list[dict[str, Any]],
) -> dict[str, str | None]:
    """Get prompt text for speech or image prompts.

    Args:
        tag_name: Either SPEECH_PROMPT or IMAGE_PROMPT
        extracted_fields: List of extracted field dicts

    Returns:
        Dict with 'prompt' and optionally 'instructions'
    """
    if tag_name == SPEECH_PROMPT:
        speech_field = next((f for f in extracted_fields if f["name"] == SPEECH_PROMPT), None)
        system_field = next((f for f in extracted_fields if f["name"] == SYSTEM), None)

        return {
            "prompt": speech_field["content"] if speech_field else "",
            "instructions": system_field["content"] if system_field else None,
        }

    if tag_name == IMAGE_PROMPT:
        image_field = next((f for f in extracted_fields if f["name"] == IMAGE_PROMPT), None)
        return {"prompt": image_field["content"] if image_field else ""}

    return {"prompt": ""}


def _build_text_config(
    name: str,
    extracted_fields: list[dict[str, Any]],
    text_settings: dict[str, Any],
    test_settings: dict[str, Any] | None,
    agentmark_meta: dict[str, Any] | None,
) -> TextConfigSchema:
    """Build a text config from extracted data."""
    messages = _get_messages(extracted_fields, "text")

    config_dict: dict[str, Any] = {
        "name": name,
        "messages": messages,
        "text_config": text_settings,
    }

    if test_settings:
        config_dict["test_settings"] = test_settings
    if agentmark_meta:
        config_dict["agentmark_meta"] = agentmark_meta

    return TextConfigSchema.model_validate(config_dict)


def _build_object_config(
    name: str,
    extracted_fields: list[dict[str, Any]],
    object_settings: dict[str, Any],
    test_settings: dict[str, Any] | None,
    agentmark_meta: dict[str, Any] | None,
) -> ObjectConfigSchema:
    """Build an object config from extracted data."""
    messages = _get_messages(extracted_fields, "object")

    config_dict: dict[str, Any] = {
        "name": name,
        "messages": messages,
        "object_config": object_settings,
    }

    if test_settings:
        config_dict["test_settings"] = test_settings
    if agentmark_meta:
        config_dict["agentmark_meta"] = agentmark_meta

    return ObjectConfigSchema.model_validate(config_dict)


def _build_image_config(
    name: str,
    extracted_fields: list[dict[str, Any]],
    image_settings: dict[str, Any],
    test_settings: dict[str, Any] | None,
    agentmark_meta: dict[str, Any] | None,
) -> ImageConfigSchema:
    """Build an image config from extracted data."""
    prompt_data = _get_prompt(IMAGE_PROMPT, extracted_fields)

    config_dict: dict[str, Any] = {
        "name": name,
        "image_config": {
            **image_settings,
            "prompt": prompt_data["prompt"],
        },
    }

    if test_settings:
        config_dict["test_settings"] = test_settings
    if agentmark_meta:
        config_dict["agentmark_meta"] = agentmark_meta

    return ImageConfigSchema.model_validate(config_dict)


def _build_speech_config(
    name: str,
    extracted_fields: list[dict[str, Any]],
    speech_settings: dict[str, Any],
    test_settings: dict[str, Any] | None,
    agentmark_meta: dict[str, Any] | None,
) -> SpeechConfigSchema:
    """Build a speech config from extracted data."""
    prompt_data = _get_prompt(SPEECH_PROMPT, extracted_fields)

    speech_config: dict[str, Any] = {
        **speech_settings,
        "text": prompt_data["prompt"],
    }
    if prompt_data.get("instructions"):
        speech_config["instructions"] = prompt_data["instructions"]

    config_dict: dict[str, Any] = {
        "name": name,
        "speech_config": speech_config,
    }

    if test_settings:
        config_dict["test_settings"] = test_settings
    if agentmark_meta:
        config_dict["agentmark_meta"] = agentmark_meta

    return SpeechConfigSchema.model_validate(config_dict)

File: packages/prompt-core/src/template_engines/templatedx.ts
import type { Ast } from "@agentmark-ai/templatedx";
import type {
  TemplateEngine,
  JSONObject,
  TestSettings,
  PromptKind,
  ImageSettings,
  ObjectSettings,
  TextSettings,
  SpeechSettings,
  RichChatMessage,
  AgentmarkConfig,
  ImagePart,
  TextPart,
  FilePart,
} from "../types";
import { 
  languageTemplateDX, 
  getTemplateDXInstance,
  determinePromptType,
  USER,
  SYSTEM,
  ASSISTANT,
  SPEECH_PROMPT,
  IMAGE_PROMPT,
} from "./templatedx-instances";

type ExtractedField = {
  name: string;
  content: string | Array<TextPart | ImagePart | FilePart>;
};

type SharedContext = {
  "__agentmark-extractTextPromises"?: Promise<ExtractedField>[];
};

const ROLETAGS = new Set([USER, SYSTEM, ASSISTANT]);

export class TemplateDXTemplateEngine implements TemplateEngine {
  async compile<
    R = AgentmarkConfig,
    P extends Record<string, any> = JSONObject
  >(options: { template: Ast; props?: P }): Promise<R> {
    return getRawConfig({ ...options, ast: options.template }) as R;
  }
}

function getMessages({
  extractedFields,
  configType,
}: {
  extractedFields: ExtractedField[];
  configType: PromptKind;
}): RichChatMessage[] {
  const messages: RichChatMessage[] = [];
  
  extractedFields.forEach((field, index) => {
    const fieldName = field.name;
    
    if (index !== 0 && fieldName === SYSTEM) {
      throw new Error(`System message may only be the first message: ${field.content}`);
    }
    
    if (!ROLETAGS.has(fieldName)) {
      throw new Error(`Invalid role tag:"${fieldName}" in config type: ${configType}.`);
    }

    if (fieldName === USER) {
      messages.push({ 
        role: "user", 
        content: field.content as any
      });
    } else if (fieldName === ASSISTANT) {
      messages.push({ 
        role: "assistant", 
        content: field.content as string 
      });
    } else if (fieldName === SYSTEM) {
      messages.push({ 
        role: "system", 
        content: field.content as string 
      });
    }
  });
  
  return messages;
}

function getPrompt({
  tagName,
  extractedFields,
}: {
  tagName: typeof SPEECH_PROMPT | typeof IMAGE_PROMPT;
  extractedFields: ExtractedField[];
}): { prompt: string; instructions?: string } {
  if (tagName === SPEECH_PROMPT) {
    const speechField = extractedFields.find((field) => field.name === SPEECH_PROMPT);
    const systemField = extractedFields.find((field) => field.name === SYSTEM);

    return {
      prompt: (speechField?.content as string) ?? "",
      instructions: systemField?.content as string,
    };
  } 
  
  if (tagName === IMAGE_PROMPT) {
    const imageField = extractedFields.find((field) => field.name === IMAGE_PROMPT);
    return { prompt: (imageField?.content as string) ?? "" };
  }
  
  return { prompt: "" };
}

export async function getRawConfig({
  ast,
  props,
}: {
  ast: Ast;
  props?: JSONObject;
}): Promise<AgentmarkConfig> {
  const frontMatter: any = languageTemplateDX.getFrontMatter(ast);
  const promptType = determinePromptType(frontMatter);
  
  const templateDXInstance = getTemplateDXInstance(promptType);
  const shared: SharedContext = {};
  await templateDXInstance.transform(ast, props || {}, shared);
  
  const extractedFieldPromises = shared["__agentmark-extractTextPromises"] || [];
  const extractedFields = await Promise.all(extractedFieldPromises);

  const name: string = frontMatter.name;
  const speechSettings: SpeechSettings | undefined = frontMatter.speech_config;
  const imageSettings: ImageSettings | undefined = frontMatter.image_config;
  const objectSettings: ObjectSettings | undefined = frontMatter.object_config;
  const textSettings: TextSettings | undefined = frontMatter.text_config;
  const testSettings: TestSettings | undefined = frontMatter.test_settings;
  const agentmarkMeta: Record<string, any> | undefined = frontMatter.agentmark_meta;

  let configType: PromptKind | undefined;
  if (speechSettings) configType = "speech";
  else if (imageSettings) configType = "image";
  else if (objectSettings) configType = "object";
  else if (textSettings) configType = "text";

  if (configType === "speech" || configType === "image") {
    const { prompt, instructions } = getPrompt({
      tagName: configType === "speech" ? SPEECH_PROMPT : IMAGE_PROMPT,
      extractedFields,
    });

    if (configType === "speech" && speechSettings && prompt) {
      return {
        name,
        speech_config: {
          ...speechSettings,
          text: prompt,
          instructions: instructions ?? "",
        },
        test_settings: testSettings,
        ...(agentmarkMeta && { agentmark_meta: agentmarkMeta }),
      };
    }
    
    if (configType === "image" && imageSettings && prompt) {
      return {
        name,
        image_config: {
          ...imageSettings,
          prompt: prompt,
        },
        test_settings: testSettings,
        ...(agentmarkMeta && { agentmark_meta: agentmarkMeta }),
      };
    }
  } else if (configType === "object" || configType === "text") {
    const messages = getMessages({ extractedFields, configType });

    if (configType === "object" && objectSettings) {
      return {
        name,
        messages,
        object_config: objectSettings,
        test_settings: testSettings,
        ...(agentmarkMeta && { agentmark_meta: agentmarkMeta }),
      };
    }
    
    if (configType === "text" && textSettings) {
      return {
        name,
        messages,
        text_config: textSettings,
        test_settings: testSettings,
        ...(agentmarkMeta && { agentmark_meta: agentmarkMeta }),
      };
    }
  }

  throw new Error("No valid config found in frontmatter.");
}
